File: packages/engine/src/template.ts
/**
 * Template engine for node configs and tool inputs. Substitutes
 * `{{context.<nodeId>.output.<field>}}`, `{{inputs.<field>}}`,
 * `{{secret.NAME}}`, and `{{env.NAME}}` references. Tracks every resolved
 * secret AND env value (length >= 4) in a redaction list so
 * `execute-node.ts` can strip plaintext from outputs and thrown errors before
 * persistence. Env vars are common carriers for API keys (e.g.
 * `OPENAI_API_KEY`, `JANUSLY_API_SERVICE_TOKEN`), so they get the same
 * scrub treatment as the explicit `{{secret.*}}` channel.
 *
 * Used by `execute-node.ts` (every node-config render),
 * `node-registry.ts:loop` (per-iteration mapping), and tool executors
 * that template their inputs.
 *
 * Invariants:
 * - The renderer never persists resolved values — it returns the rendered
 *   payload + a list of values to redact. The caller is responsible for
 *   applying `redactValues` before any DB write.
 * - Empty / missing path resolves to `""` (string interpolation), not the
 *   literal `undefined` — this is the contract `apps/web` relies on for
 *   user-friendly templating.
 */

import { redactValues } from "@janusly/shared/src/safe-persist";

import { getSecret } from "./secrets";

export type TemplateScope = Record<string, unknown>;

export const MAX_RECORDED_UNRESOLVED_PATHS = 20;

/**
 * Stable failure envelope for strict template policy. Both the dispatcher and
 * executors with late-bound scopes use this class so callers see one contract
 * regardless of when a template can first be resolved.
 */
export class UnresolvedTemplatePathError extends Error {
  readonly code = "UNRESOLVED_TEMPLATE_PATH";
  readonly count: number;
  readonly paths: string[];
  readonly truncated: boolean;

  constructor(paths: string[]) {
    super(`Node config contains ${paths.length} unresolved template path${paths.length === 1 ? "" : "s"}`);
    this.name = "UnresolvedTemplatePathError";
    this.count = paths.length;
    this.paths = paths.slice(0, MAX_RECORDED_UNRESOLVED_PATHS);
    this.truncated = paths.length > this.paths.length;
  }
}

function isTemplateRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isPathContainer(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

/**
 * Detect a `ReadableStream`-shaped value. Streams cannot be templated into
 * strings — interpolating one would either coerce to `[object ReadableStream]`
 * (useless) or surface as an opaque error downstream. We throw with the
 * resolved path so the operator can spot which `{{...}}` reference is
 * pointing at a stream-typed context value.
 *
 * The streaming-mode HTTP executor always pre-consumes into a preview before
 * returning, so context values reachable from a template should never be a
 * live stream in practice. This guard is defense-in-depth for the rare path
 * where a stream escapes its executor invocation.
 */
function isReadableStreamLike(value: unknown): boolean {
  if (!value || typeof value !== 'object') return false;
  if (typeof (globalThis as { ReadableStream?: unknown }).ReadableStream !== 'undefined'
    && value instanceof (globalThis as { ReadableStream: { new (): unknown } }).ReadableStream) {
    return true;
  }
  const v = value as { getReader?: unknown; tee?: unknown };
  return typeof v.getReader === 'function' && typeof v.tee === 'function';
}

/**
 * Thrown when a template / expression / mapping resolves a `{{...}}` path
 * to a `ReadableStream`. The `.path` field carries the offending dotted
 * reference so the operator sees exactly which node output is the source.
 */
export class StreamValueInTemplateError extends Error {
  readonly path: string;
  constructor(path: string) {
    super(`Refusing to template a ReadableStream value at path ${path}; streams must be consumed within their executor (see consumeStreamToPreview).`);
    this.name = 'StreamValueInTemplateError';
    this.path = path;
  }
}

type PathResolution = { found: boolean; value: unknown };

function resolveByPath(source: unknown, path: string): PathResolution {
  let resolved = source;
  for (const key of path.split('.')) {
    if (!isPathContainer(resolved) || !Object.prototype.hasOwnProperty.call(resolved, key)) {
      return { found: false, value: undefined };
    }
    resolved = resolved[key];
  }
  if (isReadableStreamLike(resolved)) {
    throw new StreamValueInTemplateError(path);
  }
  return { found: true, value: resolved };
}

/** Walk a dotted path through `source`. Returns `undefined` on any absent link. Throws `StreamValueInTemplateError` if the terminal value is a `ReadableStream`. */
export function getByPath(source: unknown, path: string): unknown {
  return resolveByPath(source, path).value;
}

function safeUnresolvedPath(path: string): string {
  if (path.startsWith('secret.')) return 'secret.*';
  if (path.startsWith('env.')) return 'env.*';
  return path.slice(0, 160);
}

function trackUnresolved(unresolvedPaths: Set<string>, path: string): void {
  unresolvedPaths.add(safeUnresolvedPath(path));
}

function isDeferredPath(path: string, deferredRoots: ReadonlySet<string>): boolean {
  return deferredRoots.has(path.split('.', 1)[0] ?? '');
}

function renderTemplateInternal(
  value: unknown,
  scope: TemplateScope,
  redactionList: Set<string>,
  unresolvedPaths: Set<string>,
  deferredRoots: ReadonlySet<string>,
): unknown {
  if (typeof value === 'string') {
    // Single-template-reference shape (entire string is one `{{...}}`):
    // return the resolved value as-is so arrays/objects/numbers/booleans
    // survive intact. Without this, `loop.items: "{{context.x.items}}"`
    // would receive a JSON-stringified array (then degrade via CSV-split).
    // Multi-reference strings ("hello {{name}}") still flow through the
    // string-substitution branch below.
    const singleRef = value.match(/^\s*\{\{\s*([^}]+)\s*\}\}\s*$/);
    if (singleRef) {
      const expr = singleRef[1].trim();
      if (isDeferredPath(expr, deferredRoots)) return value;
      if (expr.startsWith('secret.')) {
        const secretName = expr.replace('secret.', '').toUpperCase();
        const resolved = getSecret(secretName);
        if (resolved && resolved.length >= 4) redactionList.add(resolved);
        if (!resolved) trackUnresolved(unresolvedPaths, expr);
        return resolved;
      }
      if (expr.startsWith('env.')) {
        const envName = expr.replace('env.', '').toUpperCase();
        const envValue = process.env[envName];
        const resolved = envValue ?? '';
        if (resolved && resolved.length >= 4) redactionList.add(resolved);
        if (envValue === undefined) trackUnresolved(unresolvedPaths, expr);
        return resolved;
      }
      const resolution = resolveByPath(scope, expr);
      if (!resolution.found) {
        trackUnresolved(unresolvedPaths, expr);
        return '';
      }
      return resolution.value ?? '';
    }

    return value.replace(/{{\s*([^}]+)\s*}}/g, (match, rawPath) => {
      const expr = String(rawPath).trim();

      if (isDeferredPath(expr, deferredRoots)) return match;

      if (expr.startsWith('secret.')) {
        const secretName = expr.replace('secret.', '').toUpperCase();
        const resolved = getSecret(secretName);
        if (resolved && resolved.length >= 4) redactionList.add(resolved);
        if (!resolved) trackUnresolved(unresolvedPaths, expr);
        return resolved;
      }

      if (expr.startsWith('env.')) {
        const envName = expr.replace('env.', '').toUpperCase();
        const envValue = process.env[envName];
        const resolved = envValue ?? '';
        // Treat env values like secrets: any sufficiently long resolved value
        // is added to the redaction list so it gets scrubbed from outputs
        // before persistence. The 4-char floor avoids over-redacting toy
        // values like "0", "1", "dev", or "prod" — short non-secrets render
        // verbatim. Same length floor as the secret branch above.
        if (resolved && resolved.length >= 4) redactionList.add(resolved);
        if (envValue === undefined) trackUnresolved(unresolvedPaths, expr);
        return resolved;
      }

      const resolution = resolveByPath(scope, expr);

      if (!resolution.found) {
        trackUnresolved(unresolvedPaths, expr);
        return '';
      }
      const resolved = resolution.value;
      if (resolved == null) return '';
      if (typeof resolved === 'object') return JSON.stringify(resolved);
      return String(resolved);
    });
  }

  if (Array.isArray(value)) {
    return value.map((item) => renderTemplateInternal(item, scope, redactionList, unresolvedPaths, deferredRoots));
  }

  if (isTemplateRecord(value)) {
    return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, renderTemplateInternal(item, scope, redactionList, unresolvedPaths, deferredRoots)]));
  }

  return value;
}

/**
 * Render a value's template strings without tracking redactions.
 *
 * Overloads honestly express the return type:
 * - **string input** can return `unknown` — a single-template-reference
 *   string (`"{{x}}"`) returns the resolved value's native type, which
 *   could be a number, array, object, or boolean. Callers must narrow.
 * - **object / array / primitive input** preserves its structural shape.
 *   Nested single-ref strings inside the structure return raw values too,
 *   but at the top level the return type matches the input type.
 */
export function renderTemplate(value: string, scope: TemplateScope): unknown;
export function renderTemplate<T>(value: T, scope: TemplateScope): T;
export function renderTemplate<T>(value: T, scope: TemplateScope): T | unknown {
  return renderTemplateInternal(value, scope, new Set(), new Set(), new Set()) as T;
}

/**
 * Render a value's template strings AND collect every resolved
 * `{{secret.NAME}}` / `{{env.NAME}}` value into `redactedValues`. The
 * caller passes that list to `redactValues` / `redactError` before any
 * executor result or failure is persisted.
 */
export function renderTemplateWithRedactions(
  value: unknown,
  scope: TemplateScope,
  options: { deferredRoots?: readonly string[] } = {},
): { rendered: unknown; redactedValues: string[]; unresolvedPaths: string[] } {
  const list = new Set<string>();
  const unresolvedPaths = new Set<string>();
  const rendered = renderTemplateInternal(
    value,
    scope,
    list,
    unresolvedPaths,
    new Set(options.deferredRoots ?? []),
  );
  return { rendered, redactedValues: Array.from(list), unresolvedPaths: Array.from(unresolvedPaths) };
}

/**
 * Recursively replace any string occurrences of `redactedValues` (e.g.
 * resolved secret/env values) with the literal `"[redacted]"`. Applied to
 * executor outputs and `waiting` metadata before persistence.
 *
 * Implementation lives in `@janusly/shared/src/safe-persist` (single
 * source shared with `safePersistPayload` and the data layer); re-exported
 * here so existing engine imports keep working.
 */
export { redactValues };

/**
 * Redact resolved secret/env values from thrown errors before the runtime
 * serializes them into retry events, node failures, or DLQ rows. `Error`
 * fields such as `message` and `stack` are non-enumerable, so the generic
 * object walk in `redactValues` cannot handle them by itself.
 */
export function redactError(error: unknown, redactedValues: string[]): unknown {
  if (redactedValues.length === 0) return error;
  if (error instanceof Error) {
    error.message = redactValues(error.message, redactedValues);
    if (error.stack) error.stack = redactValues(error.stack, redactedValues);

    const errorRecord = error as Error & {
      cause?: unknown;
      code?: unknown;
      statusCode?: unknown;
    };
    if (errorRecord.cause !== undefined) errorRecord.cause = redactValues(errorRecord.cause, redactedValues);
    if (errorRecord.code !== undefined) errorRecord.code = redactValues(errorRecord.code, redactedValues);
    return error;
  }
  return redactValues(error, redactedValues);
}

/** Render a mapping object/string against `scope`. Used by `transform` / `loop` / tool inputs. */
export function mapInput<T>(mapping: T | null | undefined, scope: TemplateScope): T | Record<string, never> {
  return renderTemplate((mapping ?? {}) as T | Record<string, never>, scope);
}

File: internal/grammar/template.go
// Template engine for node configs and tool inputs, implements the
// reference: substitutes {{context.<nodeId>.output.<path>}},
// {{context.input.<name>}}, {{inputs.<name>}}, {{secret.NAME}} and
// {{env.NAME}}. Every resolved secret AND env value (length >= 4) lands in a
// redaction list so callers can strip plaintext from outputs and errors
// before persistence — the renderer itself never persists anything.
//
// Contract inherited from the contract:
//   - a string that is exactly one {{...}} reference returns the resolved
//     value's native type (arrays/objects/numbers survive);
//   - multi-reference strings interpolate (objects as JSON, null as "");
//   - a missing ordinary path renders "" and is tracked as unresolved;
//   - a missing env renders "" (tracked as env.*); a missing secret is a
//     hard failure.
package grammar

import (
	"encoding/json"
	"fmt"
	"regexp"
	"slices"
	"strings"
)

// MaxRecordedUnresolvedPaths bounds the paths carried by the strict-policy
// error envelope.
const MaxRecordedUnresolvedPaths = 20

var singleRefPattern = regexp.MustCompile(`^\s*\{\{\s*([^}]+?)\s*\}\}\s*$`)
var multiRefPattern = regexp.MustCompile(`\{\{\s*([^}]+?)\s*\}\}`)

// WholeTemplateReference reports the expression carried by a string whose
// entire value is exactly one Janusly template reference. Whole references
// are special because rendering preserves the referenced value's native JSON
// type; interpolated references always render as strings.
func WholeTemplateReference(value string) (expression string, ok bool) {
	match := singleRefPattern.FindStringSubmatch(value)
	if match == nil {
		return "", false
	}
	expression = strings.TrimSpace(match[1])
	if expression == "" {
		return "", false
	}
	return expression, true
}

// RenderOptions injects the environment and secret sources plus the
// executor-owned roots whose resolution is deferred to a later lifecycle
// point (loop item/index, sequential previousAgents).
type RenderOptions struct {
	// LookupEnv resolves an environment variable; nil disables env access
	// (every env reference then reads as unset). Mirrors os.LookupEnv.
	LookupEnv func(name string) (string, bool)
	// LookupSecret resolves a secret by UPPERCASED name; a false return is
	// the hard "Missing secret" failure. Nil behaves as all-missing.
	LookupSecret func(name string) (string, bool)
	// DeferredRoots are path roots left verbatim for a later render pass.
	DeferredRoots []string
}

// RenderResult carries the rendered value plus the redaction and
// unresolved-path evidence the caller must act on before persisting.
type RenderResult struct {
	Rendered        any
	RedactedValues  []string
	UnresolvedPaths []string
}

// MissingSecretError is the hard failure for an unresolvable {{secret.X}};
// unlike ordinary paths it never degrades to an empty string.
type MissingSecretError struct{ Name string }

func (e *MissingSecretError) Error() string { return "Missing secret: " + e.Name }

// UnresolvedTemplatePathError is the stable strict-policy envelope: count,
// bounded path list, truncation marker. Emitted by callers when a persisted
// templatePolicy of "strict" turns unresolved evidence into a failure.
type UnresolvedTemplatePathError struct {
	Count     int
	Paths     []string
	Truncated bool
}

func (e *UnresolvedTemplatePathError) Error() string {
	plural := "s"
	if e.Count == 1 {
		plural = ""
	}
	return fmt.Sprintf("Node config contains %d unresolved template path%s", e.Count, plural)
}

func (*UnresolvedTemplatePathError) ErrorName() string    { return "UnresolvedTemplatePathError" }
func (*UnresolvedTemplatePathError) ErrorCode() string    { return "UNRESOLVED_TEMPLATE_PATH" }
func (*UnresolvedTemplatePathError) ErrorStatusCode() int { return 0 }

// NewUnresolvedTemplatePathError bounds the recorded paths like the
// reference envelope.
func NewUnresolvedTemplatePathError(paths []string) *UnresolvedTemplatePathError {
	bounded := paths
	if len(bounded) > MaxRecordedUnresolvedPaths {
		bounded = bounded[:MaxRecordedUnresolvedPaths]
	}
	return &UnresolvedTemplatePathError{
		Count: len(paths), Paths: bounded, Truncated: len(paths) > len(bounded),
	}
}

type renderState struct {
	opts          RenderOptions
	redacted      []string
	redactedSeen  map[string]bool
	unresolved    []string
	unresolvedSet map[string]bool
	deferredRoots map[string]bool
}

// RenderTemplate renders without redaction tracking; a lone {{...}} string
// returns the resolved value's native type, so callers must narrow.
func RenderTemplate(value any, scope map[string]any) (any, error) {
	result, err := RenderTemplateWithRedactions(value, scope, RenderOptions{})
	if err != nil {
		return nil, err
	}
	return result.Rendered, nil
}

// RenderTemplateWithRedactions renders value against scope and collects
// every resolved secret/env value into the redaction list, plus every
// unresolved ordinary path (secret./env. names are anonymized).
func RenderTemplateWithRedactions(value any, scope map[string]any, opts RenderOptions) (RenderResult, error) {
	state := &renderState{
		opts:          opts,
		redactedSeen:  map[string]bool{},
		unresolvedSet: map[string]bool{},
		deferredRoots: map[string]bool{},
	}
	for _, root := range opts.DeferredRoots {
		state.deferredRoots[root] = true
	}
	rendered, err := renderValue(value, scope, state)
	if err != nil {
		return RenderResult{}, err
	}
	// Go maps do not retain the authored JSON property order. Canonicalize the
	// evidence once after rendering so retries and replicas emit the same bounded
	// path list rather than inheriting randomized map iteration.
	slices.Sort(state.unresolved)
	return RenderResult{
		Rendered:        rendered,
		RedactedValues:  state.redacted,
		UnresolvedPaths: state.unresolved,
	}, nil
}

// MapInput renders a mapping object/string against scope; nil maps to {}.
// Used by transform, loop iterations, and tool-input templating.
func MapInput(mapping any, scope map[string]any) (any, error) {
	if mapping == nil {
		return map[string]any{}, nil
	}
	return RenderTemplate(mapping, scope)
}

// GetByPath walks a dotted path (numeric segments index arrays); any absent
// link returns (nil, false), mirroring the contract's undefined.
func GetByPath(source any, path string) (any, bool) {
	current := source
	for key := range strings.SplitSeq(path, ".") {
		child := childValue(current, key)
		if isUndefined(child) {
			return nil, false
		}
		current = child
	}
	return current, true
}

func renderValue(value any, scope map[string]any, state *renderState) (any, error) {
	switch v := value.(type) {
	case string:
		return renderString(v, scope, state)
	case []any:
		out := make([]any, len(v))
		for i, item := range v {
			rendered, err := renderValue(item, scope, state)
			if err != nil {
				return nil, err
			}
			out[i] = rendered
		}
		return out, nil
	case map[string]any:
		out := make(map[string]any, len(v))
		for key, item := range v {
			rendered, err := renderValue(item, scope, state)
			if err != nil {
				return nil, err
			}
			out[key] = rendered
		}
		return out, nil
	default:
		return value, nil
	}
}

func renderString(value string, scope map[string]any, state *renderState) (any, error) {
	// Single-template-contract shape: the whole string is one {{...}}, so
	// the resolved value keeps its native type — without this, a templated
	// array would degrade into its JSON string form.
	if expr, wholeReference := WholeTemplateReference(value); wholeReference {
		if state.deferredRoots[pathRoot(expr)] {
			return value, nil
		}
		if strings.HasPrefix(expr, "secret.") {
			return resolveSecret(expr, state)
		}
		if strings.HasPrefix(expr, "env.") {
			return resolveEnv(expr, state), nil
		}
		resolved, found := GetByPath(scope, expr)
		if !found {
			state.trackUnresolved(expr)
			return "", nil
		}
		if resolved == nil {
			return "", nil
		}
		return resolved, nil
	}

	var firstErr error
	rendered := multiRefPattern.ReplaceAllStringFunc(value, func(match string) string {
		if firstErr != nil {
			return match
		}
		expr := strings.TrimSpace(multiRefPattern.FindStringSubmatch(match)[1])
		if state.deferredRoots[pathRoot(expr)] {
			return match
		}
		if strings.HasPrefix(expr, "secret.") {
			resolved, err := resolveSecret(expr, state)
			if err != nil {
				firstErr = err
				return match
			}
			return resolved.(string)
		}
		if strings.HasPrefix(expr, "env.") {
			return resolveEnv(expr, state).(string)
		}
		resolved, found := GetByPath(scope, expr)
		if !found {
			state.trackUnresolved(expr)
			return ""
		}
		if resolved == nil {
			return ""
		}
		switch resolved.(type) {
		case map[string]any, []any:
			// Interpolated objects serialize as JSON. Go marshals map keys
			// alphabetically where JS keeps insertion order — an accepted
			// divergence recorded in the plan.
			raw, err := json.Marshal(resolved)
			if err != nil {
				return ""
			}
			return string(raw)
		default:
			return jsString(resolved)
		}
	})
	if firstErr != nil {
		return nil, firstErr
	}
	return rendered, nil
}

func resolveSecret(expr string, state *renderState) (any, error) {
	name := strings.ToUpper(strings.TrimPrefix(expr, "secret."))
	var resolved string
	ok := false
	if state.opts.LookupSecret != nil {
		resolved, ok = state.opts.LookupSecret(name)
	}
	if !ok || resolved == "" {
		return nil, &MissingSecretError{Name: name}
	}
	if len(resolved) >= 4 {
		state.trackRedacted(resolved)
	}
	return resolved, nil
}

func resolveEnv(expr string, state *renderState) any {
	name := strings.ToUpper(strings.TrimPrefix(expr, "env."))
	var value string
	defined := false
	if state.opts.LookupEnv != nil {
		value, defined = state.opts.LookupEnv(name)
	}
	if !defined {
		state.trackUnresolved(expr)
		return ""
	}
	if value != "" && len(value) >= 4 {
		state.trackRedacted(value)
	}
	return value
}

func (s *renderState) trackRedacted(value string) {
	if s.redactedSeen[value] {
		return
	}
	s.redactedSeen[value] = true
	s.redacted = append(s.redacted, value)
}

// trackUnresolved records a deduplicated, privacy-safe form of the path:
// secret/env names are anonymized to their root, ordinary paths bounded to
// 160 characters.
func (s *renderState) trackUnresolved(path string) {
	safe := path
	switch {
	case strings.HasPrefix(path, "secret."):
		safe = "secret.*"
	case strings.HasPrefix(path, "env."):
		safe = "env.*"
	case len(path) > 160:
		safe = path[:160]
	}
	if s.unresolvedSet[safe] {
		return
	}
	s.unresolvedSet[safe] = true
	s.unresolved = append(s.unresolved, safe)
}

func pathRoot(expr string) string {
	root, _, _ := strings.Cut(expr, ".")
	return root
}
